### PR TITLE
feat: implement fluent validation

### PR DESCRIPTION
**Pull request by bianca-soliz-jalafund**

### CODE REVIEW

File: src/Infraestructure/DependencyInjection.cs [hunk 1]
```diff
@@ -1,73 +1,98 @@
...
+    {
+        services.AddScoped<IPasswordHasher, PasswordHasher>();
+        services.AddScoped<IJwtService, JwtService>();
+        services.AddScoped<IAuthService, AuthService>();
+        services.AddTransient<IEmailSender, EmailSender>();
+        services.AddSingleton<SmtpOptions>();
+        return services;
+    }
+    
+    private static IServiceCollection AddValidators(this IServiceCollection services)
```
Comment (axel-ayala-jalafund) [inline src/Infraestructure/DependencyInjection.cs]: You added only this method, why you modified the entire class for this, you need to modify only this

File: src/Services/AuthService.cs [hunk 2]
```diff
@@ -1,89 +1,118 @@
...
+        _loginValidator = loginValidator;
+        _createValidator = createValidator;
+        _requestResetValidator = requestResetValidator; 
+        _resetValidator = resetValidator;
+    }
+
+    public async Task<LoginResponse> LoginUserAsync(string email, string password)
+    {
+        var validationResult = await _loginValidator.ValidateAsync( new LoginRequest(email, password) );
+        validationResult.ThrowIfValidationFailed();
```
Comment (axel-ayala-jalafund) [inline src/Services/AuthService.cs]: Same thing with this, you're only modifying theses 2 lines, but, the entire class is modified.

File: src/Services/BudgetService.cs [hunk 3]
```diff
@@ -1,122 +1,135 @@
...
-
-        budgetToUpdate.BudgetAmount = budget.BudgetAmount;
-        budgetToUpdate.AlertThreshold = budget.AlertThreshold;
-
-        await _budgetRepository.UpdateAsync(budgetToUpdate);
-
-        return budgetToUpdate;
-    }
-}
+using Domain.DTOs;
```
Comment (axel-ayala-jalafund) [inline src/Services/BudgetService.cs]: Also this class, modify just necessary lines, not all.

File: src/Services/ExpenseService.cs [hunk 4]
```diff
@@ -1,179 +1,193 @@
...
-        if (percentageSpent > currentBudget.AlertThreshold)
-        {
-            string subject = "Budget threshold surpassed - Expense Tracker";
-            string body = $"Dear {user.Name}, you have surpassed your budget threshold, please check your balance.";
-            await _emailSender.SendEmail(user.Email, subject, body);
-        }
-    }
-
-}
+using Domain.DTOs;
```
Comment (axel-ayala-jalafund) [inline src/Services/ExpenseService.cs]: Also this class, modify just necessary lines, not all.

File: src/Services/IncomeService.cs [hunk 5]
```diff
@@ -1,70 +1,79 @@
...
-        existingIncome.ThrowIfNull("Income not found");
-        
-        existingIncome = income.ToDomain(existingIncome!);
-        var success = await _incomeRepository.UpdateAsync(existingIncome);
-        success.ThrowIfOperationFailed("Failed to update income");
-
-        return existingIncome;
-    }   
-}
+using Domain.DTOs;
```
Comment (axel-ayala-jalafund) [inline src/Services/IncomeService.cs]: Also this class, modify just necessary lines, not all.

File: src/Services/UserService.cs [hunk 6]
```diff
@@ -1,39 +1,47 @@
...
-        userExists.ThrowIfNull("User not found");
-
-        var updatedUser = user.ToDomain(userExists!);
-        var success = await  _userRepository.UpdateAsync(updatedUser);
-        success.ThrowIfOperationFailed("Failed to update user");
-        
-        return updatedUser;
-    }
-}
+
```
Comment (axel-ayala-jalafund) [inline src/Services/UserService.cs]: Also this class, modify just necessary lines, not all.

File: src/Utils/Exception/ValidatorExtensions.cs [hunk 7]
```diff
@@ -1,47 +1,57 @@
...
-
-    public static void ThrowIfExists<T>(this IEnumerable<T> collection, string message)
-    {
-        if (collection.Any())
-        {
-            throw new BadRequestException(message);
-        }
-    }
-}
+using ExpenseTrackerGroup3.Exceptions;
```
Comment (axel-ayala-jalafund) [inline src/Utils/Exception/ValidatorExtensions.cs]: Also this class, modify just necessary lines, not all.

File: src/Infraestructure/DependencyInjection.cs [hunk 8]
```diff
@@ -1,73 +1,98 @@
...
+    {
+        services.AddScoped<IValidator<LoginRequest>, LoginValidator>();
```
Comment (ignacio-villarroel-jalafund) [inline src/Infraestructure/DependencyInjection.cs]: If i remember correctly the trainer said to not use dependency injection when working with validators in general, please remove this code and use the code in each service with composition, this means that each service will require have independent objects of each required validator to work properly.

File: src/Utils/Exception/ValidatorExtensions.cs [hunk 9]
```diff
@@ -44,4 +46,12 @@ public static void ThrowIfExists<T>(this IEnumerable<T> collection, string messa
             throw new BadRequestException(message);
         }
     }
+
+    public static void ThrowIfValidationFailed(this ValidationResult validateResult)
+    {
+        if (!validateResult.IsValid)
+        {
+            throw new ValidationException(validateResult.Errors);
```
Comment (axel-ayala-jalafund) [inline src/Utils/Exception/ValidatorExtensions.cs:54]: This exception is not captured into middleware, please added into them.

**Review by axel-ayala-jalafund — APPROVED** [review]
Well done!

**Pull request merged**